### PR TITLE
reject deferred when success/error callbacks throw

### PR DESCRIPTION
**Pull request by wawjr3d**
when there is an error in the success (including parse function) or error callbacks, should ajax fail?

### CODE REVIEW

**Comment by wawjr3d** [comment]
i had issues with getting the tests from master to run on local or travis. if you want the PR, i'll work on making them pass

File: backbone.nativeajax.js [hunk 1]
```diff
@@ -44,16 +45,33 @@
 
         var status = xhr.status;
         var data = getData(options.headers && options.headers.Accept, xhr);
+        var aCallbackFailed = false;
```
Comment (akre54) [inline backbone.nativeajax.js]: It doesn't look like this is actually being checked anywhere. What's its use?

File: backbone.nativeajax.js [hunk 2]
```diff
@@ -18,6 +18,7 @@
   var ajax = (function() {
     var xmlRe = /^(?:application|text)\/xml/;
     var jsonRe = /^application\/json/;
+    var noop = function() {};
```
Comment (akre54) [inline backbone.nativeajax.js]: This isn't being used anywhere. Whats its use?

File: backbone.nativeajax.js [hunk 3]
```diff
@@ -44,16 +45,33 @@
...
-          if (reject) reject(xhr);
+          try {
+            options.error && options.error(xhr, status, error);
+          } catch(e) {
+            reason = e;
+            aCallbackFailed = true;
+          }
         }
+
+        if (anyHandlerFailed) {
```
Comment (akre54) [inline backbone.nativeajax.js]: Should this be `aCallbackFailed` instead?

File: backbone.nativeajax.js [hunk 4]
```diff
@@ -44,16 +45,33 @@
...
+        var reason;
 
         // Check for validity.
         if (isValid(xhr)) {
-          if (options.success) options.success(data);
-          if (resolve) resolve(data);
+          try {
+            options.success && options.success(data);
+          } catch(e) {
+            reason = e;
```
Comment (akre54) [inline backbone.nativeajax.js]: I believe, though I haven't tested, that there shouldn't be a need for the `try..catch`. It should all get caught in the promise `catch` handler.

**Comment by akre54** [comment]
Karma was outdated. Fixing now.

**Comment by akre54** [comment]
All tests pass on master, 6 failing in your tests.

**Comment by akre54** [comment]
My Promise tests are super busted at the moment, and I'd love some help cleaning them up. 

The best test for this kind of thing is to check that `then` and `catch` are called from the returned promise, not that our stubbed `resolve` and `reject` are. There's probably a sinon promise package out there somewhere that will help.

Without looking too deeply into it, it seems that simply chaining off of the Promise should actually catch the error as expected:

```js
ajax(options).then(function(data) {
  // Should not be called
}).catch(function(e) {
  // The error from the options callback
});
```

File: backbone.nativeajax.js (hunk 1 above)
Comment (wawjr3d) [inline backbone.nativeajax.js]: aCallbackFailed and anyHandlerFailed used to be same thing. will fix

File: backbone.nativeajax.js (hunk 4 above)
Comment (wawjr3d) [inline backbone.nativeajax.js]: what i've come across is when the success or error callbacks is synchronous, line 50 in the original fails and then resolve/reject never happen. i don't see a promise catch handler that could handle an error in success or error

**Comment by wawjr3d** [comment]
agree on testing approach, will change to that

**Comment by wawjr3d** [comment]
@akre54 do you think it makes sense for the ajax promise to reject when success/error callbacks fail? in my use case, i was able to get around it by wrapping the success/error callbacks and "bubbling" them: https://github.com/wawjr3d/brisket/blob/feature/no_jquery_transport/lib/client/ClientAjax.js#L47. i made this PR because my code is awkward, however, i'm torn on whether it's the responsibility of nativeajax library. on the :+1: side of this PR, i can delete all that bubbling logic.

but by making this change, it suggests that the library will factor in callback status to the overall promise. if that's true, the library should support async callbacks too. i could probably move the code around to make that work, but do you think all that extra functionality is this library's responsibility?

**Comment by wawjr3d** [comment]
depending on your answer, i'll get to the code changes probably tomorrow

**Comment by akre54** [comment]
Thanks for asking. Now that I think about it I'm not sure it does. In the usual XHR style (as with all asynchronicity in JS) if a callback fails it does so silently, there's no way to catch its exception without wrapping it yourself. It's not the request that failed, it's your callback code that did.

That said, a failing callback shouldn't break the rest of Backbone.NativeAjax's flow. We could wrap userland code in a try...catch and rethrow the error, or say that any code that could potentially throw an error should be wrapped in try...catch in the user's code. What do you think?

**Comment by akre54** [comment]
I updated the tests to be a little better about using promises and created the branch [`callback-err`](https://github.com/akre54/Backbone.NativeAjax/tree/callback-err) to start on fixing your issue. Have a look around there and feel free to build on top of it.

**Comment by wawjr3d** [comment]
yeah that makes sense to just wrap all downstream stuff in try/catch for users who want to tie success/error to promise. i could also see just a try/finally around success/error so nativeajax never breaks but not connecting them to promise. 

if you want to go with try/catch, we're all set on application code with what's on your branch. i'll take a look at the tests though

**Comment by wawjr3d** [comment]
dropped the test updates on a branch: https://github.com/wawjr3d/Backbone.NativeAjax/commit/0f6f5d4569769d12f2c7f6254ee3b2f88e15826c

**Comment by herrernst** [comment]
I've also stumbled into this when trying to parse invalid JSON. Would you mind merging your branch [callback-err](https://github.com/akre54/Backbone.NativeAjax/tree/callback-err) or something similar? Thanks

**Comment by akre54** [comment]
@herrernst if you create a pull based off the work above I'll merge in.

**Comment by wawjr3d** [comment]
sorry i forgot to follow back up on this. i ended up working around it because i got into analysis paralysis over @akre54 points about the validity of this change. now that it's holiday time, i don't have a lot of time to work on it, so if you have time @herrernst, feel free to work on it in my place.

let me know if you can't. if not, i'll add it back to my todo list now that it affects more than me

**Comment by herrernst** [comment]
@wawjr3d If you could find the time later I would be happy if you could finish your work ;)

**Comment by wawjr3d** [comment]
closing this since i completely forgot for over a year :(

**Pull request closed without merging**